go/interpreter/lfvm/instructions_test.go

```diff
@@ -1192,6 +1192,104 @@ func TestGenericCreate_MaxInitCodeSizeIsNotCheckedBeforeShanghai(t *testing.T) {
 	}
 }
 
+func TestGenericCreate_ReportsErrors(t *testing.T) {
+
+	u64Overflow := new(uint256.Int).Add(uint256.NewInt(math.MaxUint64), uint256.NewInt(math.MaxUint64))
+	one := uint256.NewInt(1)
+
+	tests := map[string]struct {
+		offset, size  uint256.Int
+		kind          tosca.CallKind
+		revision      tosca.Revision
+		expectedError error
+	}{
+		"offset overflow": {
+			offset:        *u64Overflow,
+			size:          *one,
+			expectedError: errOverflow,
+		},
+		"size overflow": {
+			offset:        *uint256.NewInt(0),
+			size:          *u64Overflow,
+			expectedError: errOverflow,
+		},
+		"size+offset overflow": {
+			offset:        *uint256.NewInt(math.MaxUint64 - 4),
+			size:          *uint256.NewInt(5),
+			expectedError: errOverflow,
+		},
+		"not enough gas for code size": {
+			offset:        *one,
+			size:          *uint256.NewInt(33),
+			revision:      tosca.R12_Shanghai,
+			expectedError: errOutOfGas,
+		},
+		"gas not checked for max code size before shanghai": {
+			offset:        *one,
+			size:          *uint256.NewInt(31),
+			revision:      tosca.R11_Paris,
+			expectedError: nil,
+		},
+		"not enough gas for create2 init code hashing": {
+			offset:        *one,
+			size:          *one,
+			kind:          tosca.Create2,
+			expectedError: errOutOfGas,
+		},
+		"does not charge init code hashing in create": {
+			offset:        *one,
+			size:          *one,
+			kind:          tosca.Create,
+			expectedError: nil,
+		},
+	}
+
+	for name, test := range tests {
+		t.Run(name, func(t *testing.T) {
+			mockRunContext := tosca.NewMockRunContext(gomock.NewController(t))
+			mockRunContext.EXPECT().Call(gomock.Any(), gomock.Any()).Return(tosca.CallResult{}, nil).AnyTimes()
+			ctxt := getEmptyContext()
+			ctxt.context = mockRunContext
+			ctxt.params.Revision = test.revision
+			ctxt.gas = 3
+
+			ctxt.stack.push(uint256.NewInt(0)) // salt
+			ctxt.stack.push(&test.size)
+			ctxt.stack.push(&test.offset)
+			ctxt.stack.push(uint256.NewInt(0)) // value
+
+			err := genericCreate(&ctxt, test.kind)
+			if err != test.expectedError {
+				t.Errorf("unexpected err. wanted %v, got %v", test.expectedError, err)
+			}
+		})
+	}
+}
+
+func TestGenericCreate_ResultIsWrittenToStack(t *testing.T) {
+	CreatedAddress := tosca.Address([20]byte{19: 0x1})
+	for _, success := range []bool{true, false} {
+		runContext := tosca.NewMockRunContext(gomock.NewController(t))
+		runContext.EXPECT().Call(gomock.Any(), gomock.Any()).Return(tosca.CallResult{Success: success, CreatedAddress: CreatedAddress}, nil)
+		ctxt := getEmptyContext()
+		ctxt.context = runContext
+		ctxt.stack.push(uint256.NewInt(0))
+		ctxt.stack.push(uint256.NewInt(0))
+		ctxt.stack.push(uint256.NewInt(0))
+		err := genericCreate(&ctxt, tosca.Create)
+		if err != nil {
+			t.Fatalf("unexpected error: %v", err)
+		}
+		want := uint256.NewInt(0)
+		if success {
+			want = uint256.NewInt(1)
```

Comment:
The resulting `want` should be the created address, not 1. In this case, this happens to be the same, but the test code should make this link apparent.